source/statements.tex

```diff
@@ -1038,11 +1038,30 @@
 the first time control passes through its declaration; such a variable is
 considered initialized upon the completion of its initialization. If the
 initialization exits by throwing an exception, the initialization is not
-complete, so it will be tried again the next time control enters the
+complete, so it will be tried again the next time control passes through the
 declaration.
-If control enters the declaration concurrently while the variable is
-being initialized, the concurrent execution shall wait for completion
+If control passes through the declaration concurrently while the variable is
+being initialized, the concurrent execution waits for completion
 of the initialization.
+\begin{example}
+\begin{codeblock}
+void f() {
+  static int x = 2;                     // OK
+  static std::string s = "hello";       // OK; if an exception is thrown during the
+                                        // initialization of \tcode{s}, \tcode{s} remains uninitialized, and
+                                        // an attempt to initialize \tcode{s} (but not \tcode{x}) will be
+                                        // made the next time \tcode{f} is called.
+  thread_local const char* w = "world"; // OK, initializes \tcode{w} to \tcode{"world"} on the main thread.
```

Comment:
`w` is still constant-initialized. Can we show non-constant initialization here? (e.g. `thread_local const char* w = std::system_category().name();`).

Reply:
Its not required to, and yes.

Reply:
> Its not required to

I don't see why it is not required to. [[expr.const] p2.2](https://eel.is/c++draft/expr.const#2.2) and [p3.3](https://eel.is/c++draft/expr.const#13.3) seemly require that `w` is constant-initialized when the initializer is `"world"`.

Reply:
I was thinking of the rules for static initialization, which only require that dynamic initialization may (but is not required to) be omitted when the initializer is a constant expression. I wasn't aware of [expr.const] p2.2.

Reply:
> I don't see why it is not required to.

@frederick-vs-ja actually wait, I am right. It is true that e.g. `int x = 0;` in block scope is constant-initialized, but this is meaningless for block variables. Something being *constant-initialized* doesn't imply that it's constant initialized (https://eel.is/c++draft/basic.start.static#2). (pay attention to the hyphenation, `constant-initialized` and `constant initialization` are two completely different things)

Reply:
In this example we're initializing a variable with thread storage duration. So if it is constant-initialized, it has constant initialization.